### PR TITLE
Add section on the use Action vs. Closure

### PR DESCRIPTION
**Pull request by bmuschko**
Can be reviewed and merged at any time.

### CODE REVIEW

File: contents/index.adoc [hunk 1]
```diff
@@ -155,6 +155,68 @@ public class Generate extends DefaultTask {
 
 The first section of this guide talks about the <<plugin-development-plugin,Plugin Development plugin>>. As an added benefit of applying the plugin to your project, the task `validateTaskProperties` automatically checks for an existing input/output annotation for every public property define in a custom task type implementation.
 
+=== Modeling DSL-like APIs
+
+Capturing user input sometimes calls for a stronger modeling of exposed configuration elements with the goal to make the API more readable and easier to understand for consumers. For illustration let's consider the following extension exposed by a plugin. In its current form it simply exposes a "flat" list of properties for configuring the creation of a web site.
```
Comment (oehme) [inline contents/index.adoc]: I'd drop the first sentence, it's hard to follow and doesn't really say anything that isn't said better in the paragraph below.

File: contents/index.adoc [hunk 2]
```diff
@@ -155,6 +155,68 @@ public class Generate extends DefaultTask {
...
+[source,groovy]
+----
+site {
+    outputDir = file('build/mysite')
+    websiteUrl = 'http://gradle.org'
+    vcsUrl = 'https://github.com/gradle-guides/gradle-site-plugin'
+}
+----
+
+As the number of exposed properties grows, you might want to introduce a nested, more expressive structure. The following code snippet adds a new configuration block named `customData` as part of the extension. You might have noticed that it provides a stronger indication of what those properties mean. The indentation of the configuration block also serves as a natural enhancement of the existing DSL exposed by the extension.
```
Comment (oehme) [inline contents/index.adoc]: I'd drop the last sentence, it is repeating what was already said in the one before.

File: contents/index.adoc [hunk 3]
```diff
@@ -155,6 +155,68 @@ public class Generate extends DefaultTask {
...
+public class SiteExtension {
+    private final CustomData customData = new CustomData();
+    
+    public void customData(Action<? super CustomData> action) {
+        action.execute(customData);
+    }
+}
+----
+
+NOTE: When executed by Groovy in a `build.gradle` file, Gradle will automatically coerse the `Closure` parameter to the single interface method of `Action` via link:http://docs.groovy-lang.org/next/html/documentation/core-semantics.html#_calling_a_method_accepting_a_sam_type_with_a_closure[SAM type coercion].
```
Comment (oehme) [inline contents/index.adoc]: This is not correct. Groovy's auto-coercion would lead to the user having to qualify each property access with `it.`, e.g.

```
customData {
  it.websiteUrl = 'http://gradle.org'
  it.vcsUrl = 'https://github.com/gradle-guides/gradle-site-plugin'
}
```

That's why we generate a custom `Closure`-taking method for every `Action`-taking method we find on a "managed" object. This custom method uses a different delegation strategy and removes the need for `it.`.

This is only true for the first level, which often confuses users. If they add an `Action`-taking method to `CustomData`, it won't work. That's because it is created with `new` instead of through our `Instantiator`. We really need a `project.instantiate` method.

Until we fix this I'd probably not explain this technical detail at all and just say "use `Action`".

File: contents/index.adoc [hunk 4]
```diff
@@ -155,6 +155,68 @@ public class Generate extends DefaultTask {
...
+
+Please be aware that the same mechanism works for other domain objects in Gradle and is not limited to extensions. The same approach would work for custom tasks as well to give an example. A future of version of Gradle might also expose a public API for enabling arbitrary POJOs with the same capability.
```
Comment (oehme) [inline contents/index.adoc]: Let's drop the last sentence and just explain the API when we actually have it.

File: contents/index.adoc [hunk 5]
```diff
@@ -155,6 +155,68 @@ public class Generate extends DefaultTask {
...
+----
+public class CustomData {
+    private String websiteUrl;
+    private String vcsUrl;
+
+    // getter and setter methods
+}
+----
+
+In the extension, you'll need to create an instance of the `CustomData` class and a method that can delegate the captured values to the data instance. To ensure that the method can be used from a varity of languages, e.g. Groovy, Java and Kotlin, the parameter responsible for configuring the `CustomData` instance should _not_ be a `groovy.lang.Closure` as it would impose a dependency on the Groovy library. The better choice is to configure underlying data objects with the help of {api-reference}org/gradle/api/Action.html[org.gradle.api.Action]. The following example demonstrates the use of `Action` in an extension definition.
```
Comment (oehme) [inline contents/index.adoc]: This explanation assumes that the user is familiar with what a `Closure` is and that we have historical baggage in that area. Instead I'd drop the mention of `Closure` completely and just say "In order to get the block-syntax, you need a method that takes an `Action`".

**Comment by bmuschko** [comment]
> It's great we're finally explaining this :) I added some comments for simplification.

@oehme Thanks for your review! I addressed all comments. Can you please check again?

File: contents/index.adoc [hunk 6]
```diff
@@ -155,6 +155,66 @@ public class Generate extends DefaultTask {
...
+public class SiteExtension {
+    private final CustomData customData = new CustomData();
+    
+    public void customData(Action<? super CustomData> action) {
+        action.execute(customData);
+    }
+}
+----
+
+NOTE: When executed by Groovy in a `build.gradle` file, Gradle will automatically generate a custom `Closure`-taking method for every method that defines an `Action` as the last parameter in its signature. The enhancement mechanism isn't limited to extensions. It also applies to other domain objects in Gradle e.g. tasks. Please be aware that objects created with the `new` keyword are not automatically enhanced by this mechanism. In those cases users will have to explicitly use the `Action` parameter.
```
Comment (oehme) [inline contents/index.adoc]: I'd drop this technical detail completely. Once we add `project.instantiate` we can adjust this guide to use that instead of a constructor call.

Comment (bmuschko) [inline contents/index.adoc]: I kind of feel that we should explicitly explain the shortcoming here to prevent additional confusion in the meantime. Otherwise, I as a user, would take for granted that it works in all situations. And who knows when we get around to making `Instantiator` a public API. We can make it less technical if needed.

Comment (oehme) [inline contents/index.adoc]: Then I would phrase it shorter and more actionable:

"If you need second or third level nesting, you will also have to add overloads that take `Closure`, because Gradle cannot instrument nested extensions at the moment."

Comment (bmuschko) [inline contents/index.adoc]: Happy with that. Done.
